src/process-win.cpp

```diff
@@ -102,28 +106,47 @@ auto Process::get_status() const -> Status {
 }
 
 auto Process::collect_outputs() -> bool {
-    const auto handles      = std::array{pipes[1].output, pipes[2].output};
-    const auto wait_process = WaitForSingleObject(process_handle, 0);
-    if(wait_process == WAIT_OBJECT_0) {
-        status = Status::Finished;
-    }
-    for(auto i = 0; i < 2; i += 1) {
-        auto buf = std::array<char, 256>();
-        while(true) {
-            auto bytes_avail = DWORD();
-            if(PeekNamedPipe(handles[i], NULL, 0, NULL, &bytes_avail, NULL) == 0 || bytes_avail == 0) {
-                break;
-            }
+    // child threads to collect the (stdout, stderr) outputs
+    auto threads = std::array<std::thread, 2>(); // stdout, stderr
+    for(auto i = 0; i < threads.size(); i += 1) {
+        threads[i] = std::thread([this, i]() {
+            auto buf = std::array<char, 256>();
             auto len = DWORD();
-            if(ReadFile(handles[i], buf.data(), buf.size(), &len, NULL) != TRUE || len <= 0) {
-                break;
+            while(true) {
+                if(ReadFile(pipes[i + 1].output, buf.data(), buf.size(), &len, NULL) != TRUE || len <= 0) {
+                    break;
+                }
+                if(status == Status::Finished) {
+                    // finished the main process
+                    break;
+                }
+                auto callback = i == 0 ? on_stdout : on_stderr;
+                if(callback) {
+                    callback({buf.data(), size_t(len)});
+                }
             }
-            auto callback = i == 0 ? on_stdout : on_stderr;
-            if(callback) {
-                callback({buf.data(), size_t(len)});
+        });
+    }
+
+    // main process
+    while(true) {
+        const auto wait_process = WaitForSingleObject(process_handle, INFINITE);
+        if(wait_process == WAIT_OBJECT_0) {
+            status   = Status::Finished;
+            auto buf = std::array{' '};
+            auto len = DWORD();
+            assert_b(WriteFile(pipes[1].input, buf.data(), buf.size(), &len, NULL) == TRUE, "failed to write to the child process. GetLastError: ", GetLastError());
+            assert_b(WriteFile(pipes[2].input, buf.data(), buf.size(), &len, NULL) == TRUE, "failed to write to the child process. GetLastError: ", GetLastError());
+            while(true) {
+                if(threads[0].joinable() && threads[1].joinable()) {
+                    threads[0].join();
+                    threads[1].join();
+                    break;
+                }
             }
```

Comment:
what's the purpose of this loop?
this can be simply threads[0].join(); threads[1].join()

Reply:
I misunderstood that if std::thread::joinable() returns TRUE, it means ready to call the destructor and std::thread::join() calls the destructor of it's own thread.
I will fix it.
> this can be simply threads[0].join(); threads[1].join()
No need to check threads[0].joinable() and threads[1].joinable()?

Reply:
> No need to check threads[0].joinable() and threads[1].joinable()?
I think so because these threads will be launched unconditionally